### DISCUSSION TITLE
Chakra Factory and Typescript

### DISCUSSION THREAD

**Discussion by vgarmes** [chakra-ui/chakra-ui#5278] [Q&A]
Hi, 
I'm trying to use Chakra Factory with Typescript in order to create a custom component that can receive both Chakra's style props and the library framer-motion's props, so I wrote this:

```
const FadeIn = chakra(motion.div, {
  shouldForwardProp: prop => {
    // allow all props as long as they are chakra props plus 'transition' prop
    return shouldForwardProp(prop) || prop === 'transition'
  }
}) 
```

So I can use the component like this:
```
<FadeIn
    initial={{ y: 10, opacity: 0 }}
    animate={{ y: 0, opacity: 1 }}
    transition={{ duration: 0.8, delay: 0 }}
    mb={6}
 >
    {children}
</FadeIn>
```

but I get a typescript error that the values in 'transition' prop are not assignable to its type.

The only way I could make it work is type casting the component to `ComponentWithAs<'div', MotionProps>`, where `MotionProps` comes from the type definitions in framer-motion. Is this the right way to achieve this?

**Comment by lyubchev** [comment]
Did you find a solution

**Comment by spud93** [comment]
I also encountered this issue a while ago, I'm currently using styled-components with a prop containing the Chakra theme colours as workaround. I would love to not have to do that!

**Comment by salman-farooq-sh** [comment]
This can be a generic component which combines Chakra UI and Framer Motion. All credit goes to @haseeb-esketchers .

```
import { chakra, shouldForwardProp } from "@chakra-ui/react";
import {
  ForwardRefComponent,
  HTMLMotionProps,
  isValidMotionProp,
  motion,
} from "framer-motion";

const MotionBox = chakra<
  ForwardRefComponent<HTMLDivElement, HTMLMotionProps<"div">>,
  HTMLMotionProps<"div">
>(motion.div, {
  /**
   * Allow motion props and non-Chakra props to be forwarded.
   */
  shouldForwardProp: (prop) =>
    isValidMotionProp(prop) || shouldForwardProp(prop),
});
```
